File: Framework/LanguageSupport/java/MMIJava/src/Adapter/ThriftAdapterImplementation.java
// SPDX-License-Identifier: MIT
// The content of this file has been developed in the context of the MOSIM research project.
// Original author(s): Andreas Kaiser, Felix Gaisbauer

package Adapter;

import Utils.LogLevel;
import Utils.Logger;
import de.mosim.mmi.avatar.MAvatarDescription;
import de.mosim.mmi.constraints.MConstraint;
import de.mosim.mmi.core.MBoolResponse;
import de.mosim.mmi.mmu.*;
import de.mosim.mmi.register.MAdapterDescription;
import de.mosim.mmi.register.MMIAdapter;
import de.mosim.mmi.scene.*;
import org.apache.thrift.TException;

import java.nio.ByteBuffer;
import java.text.DateFormat;
import java.util.*;
import java.util.stream.Collectors;

public class ThriftAdapterImplementation implements MMIAdapter.Iface {

    /**
     * Implementation of the thrift adapter functionality
     */

    //	Basic initialization of a MMMU
    @Override
    public MBoolResponse Initialize(MAvatarDescription avatarDescription, Map<String, String> properties, String mmuID, String sessionID) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "Initialize");
        SessionData.lastAccess = System.currentTimeMillis();

        try {
            return SessionHandling.getMMUbyId(sessionID, mmuID).Initialize(avatarDescription, properties);
        } catch (Exception e) {
            MBoolResponse result = new MBoolResponse(false);
            String message = e.getMessage();
            Logger.printLog(LogLevel.L_ERROR, message);
            result.addToLogData(message);
            return result;
        }
    }

    //	Execute command of a MMU
    @Override
    public MBoolResponse AssignInstruction(MInstruction instruction, MSimulationState simulationState, String mmuID, String sessionID) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "AssignInstruction");
        SessionData.lastAccess = System.currentTimeMillis();

        try {
            return SessionHandling.getMMUbyId(sessionID, mmuID).AssignInstruction(instruction, simulationState);
        } catch (Exception e) {
            MBoolResponse result = new MBoolResponse(false);
            String message = e.getMessage();
            Logger.printLog(LogLevel.L_ERROR, message);
            result.addToLogData(message);
            return result;
        }
    }

    //	Basic do step routine which triggers the simulation update of the respective MMU
    @Override
    public MSimulationResult DoStep(double time, MSimulationState simulationState, String mmuID, String sessionID) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "DoStep");
        SessionData.lastAccess = System.currentTimeMillis();

        try {
            return SessionHandling.getMMUbyId(sessionID, mmuID).DoStep(time, simulationState);
        } catch (Exception e) {
            Logger.printLog(LogLevel.L_ERROR, e.getMessage());
            return new MSimulationResult();
        }
    }

    //	Returns constraints which are relevant for the transition
    @Override
    public List<MConstraint> GetBoundaryConstraints(MInstruction instruction, String mmuID, String sessionID) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "GetBoundaryConstraints");
        //System.out.println(" GetBoundaryConstraints");
        SessionData.lastAccess = System.currentTimeMillis();
        try {
            return SessionHandling.getMMUbyId(sessionID, mmuID).GetBoundaryConstraints(instruction);
        } catch (Exception e) {
            Logger.printLog(LogLevel.L_ERROR, e.getMessage());
            return new ArrayList<>();
        }
    }

    //	Check whether the instruction can be executed given the current state
    @Override
    public MBoolResponse CheckPrerequisites(MInstruction instruction, String mmuID, String sessionID) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "CheckPrerequisites");
        SessionData.lastAccess = System.currentTimeMillis();

        try {
            return SessionHandling.getMMUbyId(sessionID, mmuID).CheckPrerequisites(instruction);
        } catch (Exception e) {
            MBoolResponse result = new MBoolResponse(false);
            String message = e.getMessage();
            Logger.printLog(LogLevel.L_ERROR, message);
            result.addToLogData(message);
            return result;
        }
    }

    //	Method which forces the MMU to finish
    @Override
    public MBoolResponse Abort(String instructionID, String mmuID, String sessionID) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "Abort");
        SessionData.lastAccess = System.currentTimeMillis();

        try {
            return SessionHandling.getMMUbyId(sessionID, mmuID).Abort(instructionID);
        } catch (Exception e) {
            MBoolResponse result = new MBoolResponse(false);
            String message = e.getMessage();
            Logger.printLog(LogLevel.L_ERROR, message);
            result.addToLogData(message);
            return result;
        }
    }

    //	Method disposes the MMU
    @Override
    public MBoolResponse Dispose(String mmuID, String sessionID) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "Dispose");
        SessionData.lastAccess = System.currentTimeMillis();

        try {
            return SessionHandling.getMMUbyId(sessionID, mmuID).Dispose(new HashMap<>());
        } catch (Exception e) {
            MBoolResponse result = new MBoolResponse(false);
            String message = e.getMessage();
            Logger.printLog(LogLevel.L_ERROR, message);
            result.addToLogData(message);
            return result;
        }
    }

    //	Method for executing an arbitrary function (optionally)
    @Override
    public Map<String, String> ExecuteFunction(String name, Map<String, String> parameters, String mmuID, String sessionID) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "ExecuteFunction");
        SessionData.lastAccess = System.currentTimeMillis();

        try {
            return SessionHandling.getMMUbyId(sessionID, mmuID).ExecuteFunction(name, parameters);
        } catch (Exception e) {
            Logger.printLog(LogLevel.L_ERROR, e.getMessage());
            return new HashMap<>();
        }
    }

    //	Returns the status of the adapter
    @Override
    public Map<String, String> GetStatus() throws TException {
        // System.out.println("GetStatus");

        Map<String, String> temp = new HashMap<>();
        DateFormat df = DateFormat.getDateTimeInstance(DateFormat.FULL, DateFormat.LONG, Locale.ENGLISH);
        temp.put("Version", "0.1");
        Date date = new Date(SessionData.startTime);
        temp.put("Running since", df.format(date));
        temp.put("Total Sessions", Integer.toString(SessionData.sessionContents.size()));
        temp.put("Loadable MMUs", Integer.toString(SessionData.MMUDescriptions.size()));

        if (SessionData.lastAccess == 0) {
            temp.put("Last Access", "None");
        } else {
            date = new Date(SessionData.lastAccess);
            temp.put("Last Access", df.format(date));
        }
        return temp;
    }

    //	Returns the MAdapterDescription of the adapter in detail
    @Override
    public MAdapterDescription GetAdapterDescription() throws TException {
        return SessionData.adapterDescription;
    }

    //	Creates a session
    @Override
    public MBoolResponse CreateSession(String sessionID) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "CreateSession");
        SessionData.lastAccess = System.currentTimeMillis();
        MBoolResponse result = new MBoolResponse(true);

        try {
            SessionHandling.createSessionContent(sessionID);
        } catch (Exception e) {
            result.setSuccessful(false);
            String message = e.getMessage();
            Logger.printLog(LogLevel.L_ERROR, message);
            result.addToLogData(message);
            return result;
        }
        return result;
    }

    //	Closes the session
    @Override
    public MBoolResponse CloseSession(String sessionID) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "CloseSession");
        SessionData.lastAccess = System.currentTimeMillis();
        MBoolResponse result = new MBoolResponse(true);

        try {
            SessionHandling.removeSessionContent(sessionID);
        } catch (Exception e) {
            String message = e.getMessage();
            Logger.printLog(LogLevel.L_ERROR, message);
            result.addToLogData(message);
            return result;
        }
        return result;
    }

    //	Method to synchronize the scene
    @Override
    public MBoolResponse PushScene(MSceneUpdate sceneUpdates, String sessionID) {
        Logger.printLog(LogLevel.L_DEBUG, "PushScene");

        try {
            return SessionHandling.getSessionContentBySessionID(sessionID).SceneBuffer.Apply(sceneUpdates);
        } catch (Exception e) {
            MBoolResponse result = new MBoolResponse(false);
            String message = e.getMessage();
            Logger.printLog(LogLevel.L_ERROR, message);
            result.addToLogData(message);
            return result;
        }
    }

    //	Returns descriptions of all MMUs which can be loaded
    @Override
    public List<MMUDescription> GetLoadableMMUs() throws TException {
        //Logger.printLog(LogLevel.L_DEBUG,"GetLoadableMMUs");
        return SessionData.MMUDescriptions;
    }

    //	Returns all MMUs of the session
    @Override
    public List<MMUDescription> GetMMus(String sessionID) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "GetMMUs");
        SessionData.lastAccess = System.currentTimeMillis();

        try {
            AvatarContent avatarContent = SessionHandling.getAvatarContentBySessionID(sessionID);
            return SessionData.MMUDescriptions.stream().filter(mmuDescription -> avatarContent.MMUs.containsKey(mmuDescription.ID)).collect(Collectors.toList());
        } catch (Exception e) {
            Logger.printLog(LogLevel.L_ERROR, e.getMessage());
            return new ArrayList<>();
        }
    }

    //	Returns the description of the MMU
    @Override
    public MMUDescription GetDescription(String mmuID, String sessionID) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "GetDescription");
        SessionData.lastAccess = System.currentTimeMillis();

        return SessionData.MMUDescriptions.stream().filter(desc -> desc.Name.equals(mmuID)).findFirst().orElse(null);
    }

    //	Returns the whole scene
    @Override
    public List<MSceneObject> GetScene(String sessionID) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "GetScene");
        SessionData.lastAccess = System.currentTimeMillis();

        try {
            return SessionHandling.getSessionContentBySessionID(sessionID).SceneBuffer.GetSceneObjects();
        } catch (Exception e) {
            Logger.printLog(LogLevel.L_ERROR, e.getMessage());
            return new ArrayList<>();
        }
    }

    //	Returns the scene changes of the current frame
    @Override
    public MSceneUpdate GetSceneChanges(String sessionID) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "GetSceneChanges");
        SessionData.lastAccess = System.currentTimeMillis();

        try {
            return SessionHandling.getSessionContentBySessionID(sessionID).SceneBuffer.GetSceneChanges();
        } catch (Exception e) {
            Logger.printLog(LogLevel.L_ERROR, e.getMessage());
            return new MSceneUpdate();
        }
    }

    //	Method loads MMUs for the specific session
    @Override
    public HashMap<String,String> LoadMMUs(List<String> mmus, String sessionID) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "LoadMMUs");
        SessionData.lastAccess = System.currentTimeMillis();

        //To do
        HashMap<String, String> mmuInstanceMapping = new HashMap<String, String>();



        try {
            String[] arr = SessionTools.getSplittedIDs(sessionID);
            String sceneId = arr[0];
            String avatarId = arr[1];
            SessionContent sessionContent = SessionHandling.getSessionContentBySceneID(sceneId);

            for (String id : mmus) {
                //Tuple<String,MMUDescription> loadProperty = SessionData.MMULoadingProperties.stream().filter(s->s.y.ID.equals(id)).findAny().orElse(null);
                MMUDescription mmuDescription = SessionData.MMUZipEntry.get(id);
                if (mmuDescription != null) {
                    MotionModelUnitBase mmu = AdapterController.getMMUInstantiator().InstantiateMMU(mmuDescription);

                    if (mmu != null) {
                        //Assign the service access
                        mmu.ServiceAccess = sessionContent.serviceAccess;

                        //Assign the scene
                        mmu.SceneAccess = sessionContent.SceneBuffer;

                        Logger.printLog(LogLevel.L_INFO, "Loaded MMU: " + mmuDescription.getName() + " for session: " + sessionID);
                        AvatarContent avatarContent = sessionContent.AvatarContent.get(avatarId);

                        if (avatarContent == null) {
                            avatarContent = new AvatarContent(avatarId);
                            sessionContent.AvatarContent.putIfAbsent(avatarId, avatarContent);
                        }
                        avatarContent.MMUs.put(mmuDescription.getID(), mmu);
                    }
                }
            }
        } catch (Exception e) {
            String message = e.getMessage();
            Logger.printLog(LogLevel.L_ERROR, message);
            //result.addToLogData(message);
            //result.setSuccessful((false));
        }
        return mmuInstanceMapping;
    }

    //	Method creates checkpoint of the given MMU
    @Override
    public ByteBuffer CreateCheckpoint(String mmuID, String sessionID) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "CreateCheckpoint");
        SessionData.lastAccess = System.currentTimeMillis();

        try {
            return SessionHandling.getMMUbyId(sessionID, mmuID).CreateCheckpoint();
        } catch (Exception e) {
            Logger.printLog(LogLevel.L_ERROR, e.getMessage());
            return ByteBuffer.wrap(null);
        }
    }

    //	Restores the checkpoint of the given MMU
    @Override
    public MBoolResponse RestoreCheckpoint(String mmuID, String sessionID, ByteBuffer checkpointData) throws TException {
        Logger.printLog(LogLevel.L_DEBUG, "RestoreCheckpoint");
        SessionData.lastAccess = System.currentTimeMillis();

        try {
            return SessionHandling.getMMUbyId(sessionID, mmuID).RestoreCheckpoint(checkpointData);
        } catch (Exception e) {
            MBoolResponse result = new MBoolResponse(false);
            String message = e.getMessage();
            Logger.printLog(LogLevel.L_ERROR, message);
            result.addToLogData(message);
            return result;
        }
    }
}

File: Framework/LanguageSupport/cpp/MMICPP/Adapter/ThriftAdapterImplementation.cpp
// SPDX-License-Identifier: MIT
// The content of this file has been developed in the context of the MOSIM research project.
// Original author(s): Andreas Kaiser, Niclas Delfs, Stephan Adam

#include "ThriftAdapterImplementation.h"
#include "SessionData.h"
#include "SessionTools.h"
#include "SessionContent.h"
#include "SessionHandling.h"
#include "gen-cpp/MotionModelUnit.h"
#include "AdapterController.h"
#include "CPPMMUInstantiator.h"
#include "boost/exception/diagnostic_information.hpp"
#include <chrono>
#include "Utils/Logger.h"
#include "Extensions/MBoolResponseExtensions.h"

using namespace std;

void ThriftAdapterImplementation::Initialize(::MMIStandard::MBoolResponse & _return, const::MMIStandard::MAvatarDescription & avatarDescription, const std::map<std::string, std::string>& properties, const std::string & mmuID, const std::string & sessionID)
{
	Logger::printLog(L_DEBUG,"Initialize");
	SessionData::lastAccess = std::time(0);
	_return.__set_Successful(true);

	try
	{
		SessionHandling::GetMMUbyId(sessionID,mmuID).Initialize(_return, avatarDescription, properties);
	}
	catch (...)
	{		
		string message = boost::current_exception_diagnostic_information();
		Logger::printLog(L_ERROR, message);
		MBoolResponseExtensions::Update(_return, message,false);
	}
}

void ThriftAdapterImplementation::AssignInstruction(::MMIStandard::MBoolResponse & _return, const MInstruction & instruction, const MSimulationState & simulationState, const std::string & mmuID, const std::string & sessionID)
{
	Logger::printLog(L_DEBUG, "AssignInstruction");	
	SessionData::lastAccess = std::time(0);

	try
	{
		SessionHandling::GetMMUbyId(sessionID, mmuID).AssignInstruction(_return, instruction, simulationState);
	}
	catch (...)
	{
		string message = boost::current_exception_diagnostic_information();
		Logger::printLog(L_ERROR, message);
		MBoolResponseExtensions::Update(_return, message,false);
	}
}

void ThriftAdapterImplementation::DoStep(MSimulationResult & _return, const double time, const MSimulationState & simulationState, const std::string & mmuID, const std::string & sessionID)
{
	Logger::printLog(L_DEBUG, "DoStep");	
	SessionData::lastAccess = std::time(0);

	try
	{
		SessionHandling::GetMMUbyId(sessionID, mmuID).DoStep(_return, time, simulationState);
	}
	catch (...)
	{
		Logger::printLog(L_ERROR, boost::current_exception_diagnostic_information());
	}
}

void ThriftAdapterImplementation::GetBoundaryConstraints(std::vector<MConstraint>& _return, const MInstruction & instruction, const std::string & mmuID, const std::string & sessionID)
{
	Logger::printLog(L_DEBUG, "GetBoundaryConstraints");
	SessionData::lastAccess = std::time(0);

	try
	{
		SessionHandling::GetMMUbyId(sessionID, mmuID).GetBoundaryConstraints(_return, instruction);
	}
	catch (...)
	{
		Logger::printLog(L_ERROR, boost::current_exception_diagnostic_information());
	}
}

void ThriftAdapterImplementation::CheckPrerequisites(::MMIStandard::MBoolResponse & _return, const MInstruction & instruction, const std::string & mmuID, const std::string & sessionID)
{
	Logger::printLog(L_DEBUG, "CheckPrerequisites");
	SessionData::lastAccess = std::time(0);

	try
	{
		SessionHandling::GetMMUbyId(sessionID, mmuID).CheckPrerequisites(_return, instruction);
	}
	catch (...)
	{
		string message = boost::current_exception_diagnostic_information();
		Logger::printLog(L_ERROR, message);
		MBoolResponseExtensions::Update(_return, message,false);
	}
}

void ThriftAdapterImplementation::Abort(::MMIStandard::MBoolResponse & _return, const std::string & instructionID, const std::string & mmuID, const std::string & sessionID)
{
	Logger::printLog(L_DEBUG, "Abort");
	SessionData::lastAccess = std::time(0);

	try
	{
		SessionHandling::GetMMUbyId(sessionID, mmuID).Abort(_return, instructionID);
	}
	catch (...)
	{
		string message = boost::current_exception_diagnostic_information();
		Logger::printLog(L_ERROR, message);
		MBoolResponseExtensions::Update(_return, message,false);
	}
}

void ThriftAdapterImplementation::Dispose(::MMIStandard::MBoolResponse & _return, const std::string & mmuID, const std::string & sessionID)
{
	Logger::printLog(L_DEBUG, "Dispose");
	SessionData::lastAccess = std::time(0);

	try
	{
		map<string, string> parameter;
		SessionHandling::GetMMUbyId(sessionID, mmuID).Dispose(_return, parameter);
	}
	catch (...)
	{
		string message = boost::current_exception_diagnostic_information();
		Logger::printLog(L_ERROR, message);
		MBoolResponseExtensions::Update(_return, message, false);
	}
}

void ThriftAdapterImplementation::ExecuteFunction(std::map<std::string, std::string>& _return, const std::string & name, const std::map<std::string, std::string>& parameters, const std::string & mmuID, const std::string & sessionID)
{
	Logger::printLog(L_DEBUG, "ExecuteFunction");
	SessionData::lastAccess = std::time(0);

	try
	{
		SessionHandling::GetMMUbyId(sessionID, mmuID).ExecuteFunction(_return, name,parameters);
	}
	catch (...)
	{
		Logger::printLog(L_ERROR, boost::current_exception_diagnostic_information()); Logger::printLog(L_ERROR, boost::current_exception_diagnostic_information());
	}
}

//TODO check version
void ThriftAdapterImplementation::GetStatus(std::map<std::string, std::string>& _return)
{
	_return["Version"] = "0.1";
	_return["Running since"] = strtok(ctime(&SessionData::startTime), "\n");
	_return["Total Sessions"] = std::to_string(SessionData::SessionContents.size()); ;
	
	if (SessionData::lastAccess == 0)
	{
		_return["Last Access"] = "None";
	}
	else
	{
		_return["Last Access"] = strtok(ctime(&SessionData::lastAccess), "\n");
	}
	_return["Loadable MMMUs"] = std::to_string(SessionData::mmuDescriptions.size());
}

void ThriftAdapterImplementation::GetAdapterDescription(MAdapterDescription & _return)
{
	Logger::printLog(L_DEBUG, "GetAdapterDescription");
	_return = SessionData::adapterDescription;
}

void ThriftAdapterImplementation::CreateSession(::MMIStandard::MBoolResponse & _return, const std::string & sessionID)
{
	Logger::printLog(L_DEBUG, "CreateSession");
	SessionData::lastAccess = std::time(0);

	try
	{
		SessionHandling::CreateSessionContent(sessionID);

		_return.__set_Successful(true);
	}
	catch (...)
	{
		string message = boost::current_exception_diagnostic_information();
		Logger::printLog(L_ERROR, message);
		MBoolResponseExtensions::Update(_return, message, false);
	}	
}

void ThriftAdapterImplementation::CloseSession(::MMIStandard::MBoolResponse & _return, const std::string & sessionID)
{
	Logger::printLog(L_INFO, "CloseSession " + sessionID );

	try {
		SessionHandling::RemoveSessionContent(sessionID);
		_return.__set_Successful(true);	
	}
	catch (...)
	{
		string message = boost::current_exception_diagnostic_information();
		Logger::printLog(L_ERROR, message);
		MBoolResponseExtensions::Update(_return, message, false);
	}	
}

void ThriftAdapterImplementation::PushScene(::MMIStandard::MBoolResponse & _return, const::MMIStandard::MSceneUpdate & sceneUpdates, const std::string & sessionID)
{
	Logger::printLog(L_DEBUG, "PushScene");
	SessionData::lastAccess = std::time(0);

	try
	{
		SessionHandling::GetSessionContentBySessionID(SessionTools::GetSplittedIds(sessionID)[0]).sceneBuffer->Apply(_return,sceneUpdates);	
	}
	catch (...)
	{
		string message = boost::current_exception_diagnostic_information();
		Logger::printLog(L_ERROR, message);
		MBoolResponseExtensions::Update(_return, message, false);
	}
}

void ThriftAdapterImplementation::GetLoadableMMUs(std::vector<MMUDescription>& _return)
{
	//Logger::printLog(L_DEBUG, "GetLoadableMMUs");
	//SessionData::lastAccess = std::time(0);

	for (const MMUDescription &description : SessionData::mmuDescriptions)
	{
		_return.emplace_back(description);
	}
	
}

void ThriftAdapterImplementation::GetMMus(std::vector<MMUDescription>& _return, const std::string & sessionID)
{
	Logger::printLog(L_DEBUG, "GetMMUs");
	SessionData::lastAccess = std::time(0);

	try
	{
		const AvatarContent *avatarContent = &SessionHandling::GetAvatarContentBySessionID(sessionID);

		for(auto const& mmu : avatarContent->MMUs)
		{
			_return.emplace_back(SessionData::GetMMUDescription(mmu.first));
		}	
	}
	catch (...)
	{
		Logger::printLog(L_ERROR, boost::current_exception_diagnostic_information());
	}
}

void ThriftAdapterImplementation::GetDescription(MMUDescription & _return, const std::string & mmuID, const std::string & sessionID)
{
	Logger::printLog(L_DEBUG, "GetDescription");
	SessionData::lastAccess = std::time(0);
	_return =SessionData::GetMMUDescription(mmuID);
}

void ThriftAdapterImplementation::GetScene(std::vector<::MMIStandard::MSceneObject>& _return, const std::string & sessionID)
{
	Logger::printLog(L_DEBUG, "GetScene");
	SessionData::lastAccess = std::time(0);

	try
	{
		SessionHandling::GetSessionContentBySessionID(SessionTools::GetSplittedIds(sessionID)[0]).sceneBuffer->GetSceneObjects(_return);
	}
	catch (...)
	{
		Logger::printLog(L_ERROR, boost::current_exception_diagnostic_information());
	}
}

void ThriftAdapterImplementation::GetSceneChanges(::MMIStandard::MSceneUpdate & _return, const std::string & sessionID)
{
	Logger::printLog(L_DEBUG, "GetSceneChanges");

	try
	{		
		SessionData::lastAccess = std::time(0);
		SessionHandling::GetSessionContentBySessionID(SessionTools::GetSplittedIds(sessionID)[0]).sceneBuffer->GetSceneChanges(_return);
				
	}
	catch (...)
	{
		Logger::printLog(L_ERROR, boost::current_exception_diagnostic_information());
	}	
}

//TODO failure handling
void ThriftAdapterImplementation::LoadMMUs(std::map<std::string, std::string>& _return, const std::vector<std::string>& mmus, const std::string & sessionID)
{
	Logger::printLog(L_DEBUG, "LoadMMUs");
	try
	{
		std::vector<string> splittedIds = SessionTools::GetSplittedIds(sessionID);
		std::string sceneId = splittedIds[0];
		std::string avatarId = splittedIds[1];
		const SessionContent *sessionContent = &SessionHandling::GetSessionContentBySceneID(sceneId);
		SessionData::lastAccess = std::time(0);
		// save the session content of the id 
		//MBoolResponse SessionResults = SessionData::GetSessionContent(sessionID, out sessionContent);

		for (const std::string &mmuId : mmus)
		{
			unique_ptr<MotionModelUnitBaseIf> mmu = nullptr;
			auto it = SessionData::mmuPaths.find(mmuId);
			if (it != SessionData::mmuPaths.end())
			{
				CPPMMUInstantiator instantiator = AdapterController::GetMMUInstantiator();
				mmu = move(instantiator.InstantiateMMU(it->second));
			}
			if (mmu != nullptr)
			{
				mmu->serviceAccess = &sessionContent->GetServiceAccess();
				mmu->sceneAccess = &sessionContent->GetScene();

				Logger::printLog(L_INFO, "Loaded MMU : " + mmu->name + " for session: "+ sessionID );

				// insert values into map 
				_return.insert(std::pair<std::string, std::string>(mmuId, sessionID));  // added, sadam
				
				auto it = sessionContent->avatarContent.find(avatarId);
				if (it == sessionContent->avatarContent.end())
				{	
					unique_ptr avatarcontent = make_unique<AvatarContent>(avatarId);
					sessionContent->avatarContent[avatarId] = move(avatarcontent);
					avatarcontent->MMUs[mmuId] = move(mmu);
				}
				sessionContent->avatarContent[avatarId]->MMUs[mmuId] = move(mmu);					
			}
		}
	}
	catch (...)
	{
		string message = boost::current_exception_diagnostic_information();
		Logger::printLog(L_ERROR, message);
		//MBoolResponseExtensions::Update(_return, message, false);
	}

}

void ThriftAdapterImplementation::CreateCheckpoint(std::string & _return, const std::string & mmuID, const std::string & sessionID)
{
	Logger::printLog(L_DEBUG, "CreateCheckPoint");
	SessionData::lastAccess = std::time(0);

	try
	{
		SessionHandling::GetMMUbyId(sessionID, mmuID).CreateCheckpoint(_return);
	}
	catch (...)
	{
		Logger::printLog(L_ERROR, boost::current_exception_diagnostic_information());
	}
}

void ThriftAdapterImplementation::RestoreCheckpoint(::MMIStandard::MBoolResponse & _return, const std::string & mmuID, const std::string & sessionID, const std::string & checkpointData)
{
	Logger::printLog(L_DEBUG, "RestoreCheckpoint");
	SessionData::lastAccess = std::time(0);

	try
	{
		SessionHandling::GetMMUbyId(sessionID, mmuID).RestoreCheckpoint(_return, checkpointData);
	}
	catch (...)
	{
		string message = boost::current_exception_diagnostic_information();
		Logger::printLog(L_ERROR, message);
		MBoolResponseExtensions::Update(_return, message, false);
	}
}


